0015-authority.md

```diff
@@ -0,0 +1,478 @@
+
+## OnGainedAuthority / OnLostAuthority
+
+`OnGainedAuthority` and `OnLostAuthority` functions can be overridden to get informed about changes to the `HasAuthority` status. `OnGainedAuthority` gets called when `HasAuthority` switches from false to true. `OnLostAuthority` gets called when it switches to false from true
+```csharp
+public override void OnGainedAuthority()
+{
+    Debug.Log("Authority Gained :)");
+}
+
+public override void OnLostAuthority()
+{
+    Debug.Log("Authority Lost :(");
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Why is Authority a mode?
+Why is authority not something like ownership which you give to a client. The reason for that is because authority is a relation between multiple peers. In an authoritative model a set of peers pass inputs to the authoritative peer for processing.
+
+Because we are a star topology we have some technical limitations. For instance we could not have one client generate the inputs but pass it to another client which holds authority (because this would break the encapsulation principle of time explained in the NetworkTime RFC).
+
+Because of the above we will only support authority models where the owner of an object collects inputs and sends them to the authoritative peer which has to be either the owner or the server.
+
+Accepting inputs from multiple peers is something we will not support with the current proposed design. The authority mode system allows for further extension to support those modes.
+
+## Ownership vs Authority
+This RFC will introduce some rules and responsibilities for each of the two systems.
+
+IsOwner:
+- Only one peer can own a NetworkObject.
+- The `owner` of a NetworkObject can use `ServerRpcs` on its `NetworkBehaviour` by default.
+- The `owner` is responsible for collecting inputs and redirecting them to the authority.
+- The `owner` is encouraged write to `NetworkVariables` with readpermission = authority but not to other NetworkVariables.
+- When the owner peer disconnects, the object it owns might get destroyed when `DontDestroyWithOwner` is set to false.
+
+HasAuthority:
+- The `authority` should run gameplay code such as physics, movement, path finding, animation etc.
+- The `authority` should update NetworkVariables which get used by other clients to visualize the game state.
+
+IsServer:
+- The server has permission to change ownership and authority of NetworkObjects.
+- `IsServer` should be avoided for gameplay related code.
+
+## NetworkObject Authority
+
+NetworkObject will get an Authority Property and a way to query for authority with `HasAuthority`:
+```csharp
+public Authority Authority => m_Authority;
+
+public bool HasAuthority
+{
+    get
+    {
+        if (IsSpawned) // This is needed because else we don't know whether we are a server.
+        {
+            return IsOwner ? Authority == Authority.Owner : NetworkManager.IsServer && Authority == Authority.Server;
```

Comment:
This is obviously a form of syntactic sugar and I question do we really need this or should we just put this syntactic sugar into our specific implementations, as per one of your examples, `NetworkTransform`. Simply, `NetworkTransform` could have this getter-only property (of `HasAuthority()` method) on its own and we wouldn't touch anything at a lower-level-layers and wouldn't expose some extra public API we would maintain for a while.

Reply:
The motivation section explains this. If we do not implement this `NetworkTransform` cannot work with other components such a `NetworkAnimator`. `Authority` is a way for a NetworkBehaviour to communicate how it works and what network model it supports. If  each component needs to implement its own form of this concept we end up with a lot of duplicate code and lose many of the upsides mentioned in #motivation

Reply:
Where in the motivation section do you touch on these details?

Reply:
Only in my head apparently 😅. I'll rework the motivation section.